src/source/src/parser/avro_parser/schema_resolver.rs

```diff
@@ -0,0 +1,135 @@
+};
+use risingwave_common::error::{Result, RwError};
+use risingwave_connector::aws_utils::{default_conn_config, s3_client, AwsConfigV2};
+use url::Url;
+
+use crate::parser::schema_registry::Client;
+
+const AVRO_SCHEMA_LOCATION_S3_REGION: &str = "region";
+
+/// Read schema from s3 bucket.
+/// S3 file location format: <s3://bucket_name/file_name>
+pub(super) async fn read_schema_from_s3(
+    url: &Url,
+    properties: HashMap<String, String>,
+) -> Result<String> {
+    let bucket = url
+        .domain()
+        .ok_or_else(|| RwError::from(InternalError(format!("Illegal Avro schema path {}", url))))?;
+    if properties.get(AVRO_SCHEMA_LOCATION_S3_REGION).is_none() {
+        return Err(RwError::from(InvalidConfigValue {
+            config_entry: AVRO_SCHEMA_LOCATION_S3_REGION.to_string(),
+            config_value: "NONE".to_string(),
+        }));
+    }
+    let key = url.path().replace('/', "");
+    let config = AwsConfigV2::from(properties.clone());
+    let sdk_config = config.load_config(None).await;
+    let s3_client = s3_client(&sdk_config, Some(default_conn_config()));
+    let response = s3_client
+        .get_object()
+        .bucket(bucket.to_string())
+        .key(key)
+        .send()
+        .await
+        .map_err(|e| RwError::from(InternalError(e.to_string())))?;
+    let body_bytes = response.body.collect().await.map_err(|e| {
+        RwError::from(InternalError(format!(
+            "Read Avro schema file from s3 {}",
+            e
+        )))
+    })?;
+    let schema_bytes = body_bytes.into_bytes().to_vec();
+    String::from_utf8(schema_bytes)
+        .map_err(|e| RwError::from(InternalError(format!("Avro schema not valid utf8 {}", e))))
+}
+
+/// Read avro schema file from local file.For on-premise or testing.
+pub(super) fn read_schema_from_local(path: impl AsRef<Path>) -> Result<String> {
+    std::fs::read_to_string(path.as_ref()).map_err(|e| e.into())
+}
+
+/// Read avro schema file from local file.For common usage.
+pub(super) async fn read_schema_from_https(location: &Url) -> Result<String> {
+    let res = reqwest::get(location.clone()).await.map_err(|e| {
+        InvalidParameterValue(format!(
+            "failed to make request to URL: {}, err: {}",
+            location, e
+        ))
+    })?;
+    if !res.status().is_success() {
+        return Err(RwError::from(InvalidParameterValue(format!(
+            "Http request err, URL: {}, status code: {}",
+            location,
+            res.status()
+        ))));
+    }
+    let body = res
+        .bytes()
+        .await
+        .map_err(|e| InvalidParameterValue(format!("failed to read HTTP body: {}", e)))?;
+
+    String::from_utf8(body.into()).map_err(|e| {
+        RwError::from(InternalError(format!(
+            "read schema string from https failed {}",
+            e
+        )))
+    })
+}
+
+#[derive(Debug)]
+pub struct ConfluentSchemaResolver {
+    writer_schemas: SkipMap<i32, Schema>,
```

Comment:
Do we really have to use a SkipMap instead of the HashMap in the standard library?

Reply:
Yes， the `ConfluentSchemaResolver` is a member of a `Parser`, and the `parse` method of `SourceParser` only gives an immutable ref to self, and then, if we use a `RefCell` to wrap the std map, the `Parser` would break the synchronization-related requirements.

Reply:
@BugenZhao Please take a look.

Reply:
Can we make the `parse` take `&mut self`, which sounds reasonable to me?

Or, I guess using `moka::Cache` or even `RwLock<HashMap<..>>` is okay here.

Reply:
I remember that &mut self would make parser not send or sync, I will use Moka or lock instead.

Reply:
> &mut self would make parser not send or sync

This is irrelevant. 🤔

Reply:
> > &mut self would make parser not send or sync
> 
> This should be irrelevant. 🤔

But another thing can be confirmed, the `ConnectorSource` uses `Arc<SourceParserImpl>` to share the parser with the `ConnectorSourceReader`, so we can't get `&mut self`

Reply:
That's reasonable.

Reply:
> That's reasonable.

Done, please review again.